### PR TITLE
HSEARCH-5087 Add compatibility with OpenSearch 2.12.0

### PR DESCRIPTION
**Pull request by marko-bekhta**
https://hibernate.atlassian.net/browse/HSEARCH-5087
https://hibernate.atlassian.net/browse/HSEARCH-5088

Let's see if this won't be the same as with the Lucene update PR 😏 😄

### CODE REVIEW

**Review by yrodiere — APPROVED** [review]
LGTM, thanks.

@marko-bekhta would you mind backporting to 7.1? We could reasonably slip this in before the `.Final`.

**Comment by marko-bekhta** [comment]
I admit that running the tests for `opensearch-local.-*` was more of an accident rather then a conscious decision to tests vector search on all possible 2.x versions... but I guess it was good since apparently things were added to it over time and we probably should limit our vector integration to 2.9+ 🙈

**Comment by yrodiere** [comment]
> I guess it was good since apparently things were added to it over time and we probably should limit our vector integration to 2.9+ 🙈

+1 please do

File: util/internal/integrationtest/backend/elasticsearch/src/main/java/org/hibernate/search/util/impl/integrationtest/backend/elasticsearch/SearchBackendContainer.java [hunk 1]
```diff
@@ -153,6 +153,16 @@ private static GenericContainer<?> opensearch(DockerImageName dockerImageName) {
 				// See https://www.elastic.co/guide/en/elasticsearch/reference/7.17/modules-cluster.html#disk-based-shard-allocation
 				// See https://opensearch.org/docs/latest/opensearch/popular-api/#change-disk-watermarks-or-other-cluster-settings
 				.withEnv( "cluster.routing.allocation.disk.threshold_enabled", "false" );
+
+		ElasticsearchVersion version =
+				ElasticsearchVersion.of( ElasticsearchDistributionName.OPENSEARCH, dockerImageName.getVersionPart() );
+
+		if ( version.majorOptional().orElse( Integer.MIN_VALUE ) == 2
+				&& version.minor().orElse( Integer.MAX_VALUE ) > 11 ) {
+			// Note: For OpenSearch 2.12 and later, a custom password for the admin user is required to be passed to set-up and utilize demo configuration.
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply util/internal/integrationtest/backend/elasticsearch/src/main/java/org/hibernate/search/util/impl/integrationtest/backend/elasticsearch/SearchBackendContainer.java:162]: :facepalm:

Anyway... it doesn't hurt that much, I guess :)

**Comment by marko-bekhta** [comment]
simple dependency updates this week: 0 
😆 

soo it looks like there's a bug in new OpenSearch:

```http
PUT http://localhost:32771/test-float-index-000001
Content-Type: application/json

{
  "mappings": {
    "dynamic": "strict",
    "properties": {
      "field0_Float": {
        "type": "float",
        "doc_values": false
      }
    }
  }
}
```
```http
POST http://localhost:32771/test-float-index-000001/_doc/1
Content-Type: application/json

{
  "field0_Float": 123.12312
}
```
```http
POST http://localhost:32771/test-float-index-000001/_search
Content-Type: application/json

{
  "query": {
    "bool": {
      "must": {
        "constant_score": {
          "filter": {
            "terms": {
              "field0_Float": [
                123.12312
              ]
            }
          }
        }
      }
    }
  },
  "_source": false
}
```
results in:
```json
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_state_exception",
        "reason": "unexpected docvalues type NONE for field 'field0_Float' (expected one of [SORTED_NUMERIC, NUMERIC]). Re-index with correct docvalues type."
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "test-float-index-000001",
        "node": "KVxFfRZ6S6W5DORNcyZmig",
        "reason": {
          "type": "illegal_state_exception",
          "reason": "unexpected docvalues type NONE for field 'field0_Float' (expected one of [SORTED_NUMERIC, NUMERIC]). Re-index with correct docvalues type."
        }
      }
    ],
    "caused_by": {
      "type": "illegal_state_exception",
      "reason": "unexpected docvalues type NONE for field 'field0_Float' (expected one of [SORTED_NUMERIC, NUMERIC]). Re-index with correct docvalues type.",
      "caused_by": {
        "type": "illegal_state_exception",
        "reason": "unexpected docvalues type NONE for field 'field0_Float' (expected one of [SORTED_NUMERIC, NUMERIC]). Re-index with correct docvalues type."
      }
    }
  },
  "status": 500
}
```

if I change `"doc_values": false` to `"doc_values": true` it starts to work ok 😕

**Comment by yrodiere** [comment]
It's odd. I don't see why docvalues would be required here.

Maybe try to report that, and disable the corresponding test for OpenSearch 2.12 for now?

**Pull request merged**